### DISCUSSION TITLE
How to convert a FigureResampler to a json format

### DISCUSSION THREAD

**Discussion by erwanmq** [predict-idlab/plotly-resampler#283] [Q&A]
Hi!

I am using plotly_resampler with manual mode to plot hundred of thousands points in a plotly graph. I want the user to add or remove events from this plot. To add the events, it's quite simple, I have to call .add_trace(). But to remove the events 2 options are available. Either I use .visible = False or I transform the FigureResampler to a json format and remove the last data value.
But when I do this, the json format loses all the points that are not plotted in the graph. So I can lost 80-90% of my points and I can't get them back.

Is there a solution to erase a trace from a FigureResampler without losing all points and losing the state of 'FigureResampler'?

I should add that I would like to avoid .visible = False for the sake of performance. Thank you!

**Comment by jonasvdd** [comment]
Hi,

## Regarding conversion to a `json` format

`FigureResampler` only renders aggregated data of the high-frequency traces. 
This is achieved by storing this aggregated data in plotly's default *data* property and storing the raw, high-frequency, data in the (non-plotly) [_hf_data](https://github.com/predict-idlab/plotly-resampler/blob/0a68cb76e7893d19f0e5f252b81699b110f9638c/plotly_resampler/figure_resampler/figure_resampler_interface.py#L517) property.

As such, when the plotly's `to_json()` is invoked, only the aggregated data, stored within the *data* property, will be serialized into the `.json` file. So there is no convenient method to store the high frequency data into a `.json` file. Also remark that `.json` is an inefficient manner to store (large amounts of) data as e.g. floats will be encoded as strings. 

However, `FigureResampler` figures can be serialized into pickle objects (which stores the data-as-is (i.e. in binary format). Maybe you could use this format to save your `FigureResampler` figures?

https://github.com/predict-idlab/plotly-resampler/blob/0a68cb76e7893d19f0e5f252b81699b110f9638c/tests/test_serialization.py#L58-L65

## Regarding erasing a trace
I quickly glanced at the plotly-community forum, but found no straightforward solutions to remove a trace.

After toying around with this in a notebook I came to a (not 100% clean) solution.

1. Create a figure with multiple traces:
```python
import plotly.graph_objects as go
from plotly_resampler import FigureResampler

import numpy as np

labels = list(range(0, 3))
np.random.seed(0xDEADBEEF)
x = np.arange(100000)
y = np.random.normal(size=100000)

fig = FigureResampler()
fig.add_trace(go.Scatter(x=x, y=y, mode='lines', line_color='lightblue'))
fig.add_trace(go.Scatter(x=x, y=y + 10, mode='lines+markers', line_color='salmon'))
fig.add_trace(go.Scatter(x=x, y=y + 20, mode='lines', line_color='lightseagreen'))
fig.update_layout(height=500, margin=dict(l=0, r=0, t=0, b=0))
fig.show_dash()
```
![image](https://github.com/predict-idlab/plotly-resampler/assets/38005924/851e283a-e0f7-4a4e-a6d6-43d98bec002d)


2. Delete a trace from the above created figure:
```python
trace_index = 1  # specify the trace index

# delete the `hf_data` entry
fig._hf_data.pop(fig.data[trace_index].uid)  # the hf_data is a dict keyed by trace uid

#  delete the trace itself in the `data` list
fig.data[1].visible = False
fig.data[1].x = []
fig.data[1].y = []
```
![image](https://github.com/predict-idlab/plotly-resampler/assets/38005924/ae3bdb70-07bc-4436-9c49-12c39ef06868)

## Regarding removing a data-value (e.g. the last value) from a trace

Could you provide me with a minimal code example regarding your use-case?

----

I hope this helps you any further!
Kind regards,
Jonas
